core/multigrid/pgm.cpp

```diff
@@ -199,10 +210,290 @@ void Pgm<ValueType, IndexType>::generate()
 
     // Construct the coarse matrix
     // TODO: improve it
-    auto coarse_matrix = generate_coarse(exec, pgm_op, num_agg, agg_);
+    auto coarse_matrix =
+        generate_coarse(exec, local_matrix.get(), num_agg, agg_);
+
+    return std::tie(prolong_row_gather, coarse_matrix, restrict_sparsity);
+}
+
+
+#if GINKGO_BUILD_MPI
+template <typename ValueType, typename IndexType>
+template <typename GlobalIndexType>
+void Pgm<ValueType, IndexType>::communicate(
+    std::shared_ptr<const experimental::distributed::Matrix<
+        ValueType, IndexType, GlobalIndexType>>
+        matrix,
+    const array<IndexType>& local_agg, array<IndexType>& non_local_agg)
+{
+    auto exec = gko::as<LinOp>(matrix)->get_executor();
+    const auto comm = matrix->get_communicator();
+    auto send_sizes = matrix->send_sizes_;
+    auto recv_sizes = matrix->recv_sizes_;
+    auto send_offsets = matrix->send_offsets_;
+    auto recv_offsets = matrix->recv_offsets_;
+    auto gather_idxs = matrix->gather_idxs_;
+    auto total_send_size = send_offsets.back();
+    auto total_recv_size = recv_offsets.back();
+
+    array<IndexType> send_agg(exec, total_send_size);
+    exec->run(pgm::make_gather_index(
+        send_agg.get_size(), local_agg.get_const_data(),
+        gather_idxs.get_const_data(), send_agg.get_data()));
+
+    auto use_host_buffer = experimental::mpi::requires_host_buffer(exec, comm);
+    array<IndexType> host_recv_buffer(exec->get_master());
+    array<IndexType> host_send_buffer(exec->get_master());
+    if (use_host_buffer) {
+        host_recv_buffer.resize_and_reset(total_recv_size);
+        host_send_buffer.resize_and_reset(total_send_size);
+        exec->get_master()->copy_from(exec, total_send_size,
+                                      send_agg.get_data(),
+                                      host_send_buffer.get_data());
+    }
+    auto type = experimental::mpi::type_impl<IndexType>::get_type();
+
+    const auto send_ptr = use_host_buffer ? host_send_buffer.get_const_data()
+                                          : send_agg.get_const_data();
+    auto recv_ptr = use_host_buffer ? host_recv_buffer.get_data()
+                                    : non_local_agg.get_data();
+    exec->synchronize();
+    comm.all_to_all_v(use_host_buffer ? exec->get_master() : exec, send_ptr,
+                      send_sizes.data(), send_offsets.data(), type, recv_ptr,
+                      recv_sizes.data(), recv_offsets.data(), type);
+    if (use_host_buffer) {
+        exec->copy_from(exec->get_master(), total_recv_size, recv_ptr,
+                        non_local_agg.get_data());
+    }
+}
+#endif
+
+
+template <typename IndexType>
+void generate_non_local_map(
+    const std::vector<experimental::distributed::comm_index_type>& recv_offsets,
+    array<IndexType>& non_local_agg, array<IndexType>& non_local_col_map,
+    array<IndexType>& renumber)
+{
+    auto exec = renumber.get_executor();
```

Comment:
Can `renumber` be guaranteed to be on host ? Maybe it should be 
```suggestion
    auto exec = renumber.get_executor()->get_master();
```
if you want to make sure it is on the host ? 

and also probably `GKO_ASSERT(exec == exec->get_master())` in the functions where any `array` is needs to be accessed from host would be good.

Reply:
Yes, it will be on the host. I add the assert to check